app/src/org/commcare/android/util/MediaUtil.java

```diff
@@ -1,82 +1,274 @@
+            float heightScale = ((float)boundingHeight) / desiredHeight;
+            float widthScale = ((float)boundingWidth) / desiredWidth;
+            float boundingScaleDownFactor = Math.min(heightScale, widthScale);
+            desiredHeight = Math.round(desiredHeight * boundingScaleDownFactor);
+            desiredWidth = Math.round(desiredWidth * boundingScaleDownFactor);
+        }
+        try {
+            BitmapFactory.Options o = new BitmapFactory.Options();
+            o.inScaled = false;
+            Bitmap originalBitmap = BitmapFactory.decodeFile(imageFilepath, o);
+            try {
+                return Bitmap.createScaledBitmap(originalBitmap, desiredWidth, desiredHeight, false);
+            } catch (OutOfMemoryError e) {
+                return originalBitmap;
+            }
+        }
+        catch (OutOfMemoryError e) {
+            // Just inflating the image at its original size caused an OOM error, don't have a
+            // choice but to scale down
+            return performSafeScaleDown(imageFilepath, 2, 1);
+        }
+    }
+
+    public static double computeInflationScaleFactor(DisplayMetrics metrics, int targetDensity) {
+        final int SCREEN_DENSITY = metrics.densityDpi;
+
+        double actualNativeScaleFactor = metrics.density;
+        // The formula below is what Android *usually* uses to compute the value of metrics.density
+        // for a device. If this is in fact the value being used, we are not interested in it.
+        // However, if the actual value differs at all from the standard calculation, it means
+        // Android is taking other factors into consideration (such as straight up screen size),
+        // and we want to incorporate that proportionally into our own version of the scale factor
+        double standardNativeScaleFactor = (double)SCREEN_DENSITY / DisplayMetrics.DENSITY_DEFAULT;
+        double proportionalAdjustmentFactor = 1;
+        if (actualNativeScaleFactor > standardNativeScaleFactor) {
+            proportionalAdjustmentFactor = 1 +
+                    ((actualNativeScaleFactor - standardNativeScaleFactor) / standardNativeScaleFactor);
+        } else if (actualNativeScaleFactor < standardNativeScaleFactor) {
+            proportionalAdjustmentFactor = actualNativeScaleFactor / standardNativeScaleFactor;
+        }
+
+        // Get our custom scale factor, based on this device's density and what the image's target
+        // density was
+        double customDpiScaleFactor = (double)SCREEN_DENSITY / targetDensity;
+
+        return customDpiScaleFactor * proportionalAdjustmentFactor;
+    }
+
+
+    /**
+     * @return A bitmap representation of the given image file, potentially adjusted from the
+     * image's original size such that its width is no larger than containerWidth, and its height
+     * is no larger than containerHeight
+     */
+    private static Bitmap getBitmapScaledToContainer(String imageFilepath, int containerHeight,

```

Comment:
This only scales down in the BitmapFactory scale proportions, right? That seems a bit problematic. Feels like we should do one of:

1) Scale down to the _last_ scale factor that is still larger than the container (IE: the bounding box is 100x100, the image is 480x480, scale x 4 gives us 120x120, which then in an imageview is displayed as 100x100. If we scaled down x5 (the current behavior, since round(4.8) -> 5) we get 60x60, which then wouldn't fit in the container